### PR TITLE
feat(cards): #MAG-58 add linker directive

### PR DESCRIPTION
**Pull request by Sofianebernoussi**
## Describe your changes

- allow user to add a link to other application as a card
- allow user to add a link to a website as a card

## Checklist tests

- check if external linker card creation is working
- check if internal linker card creation is working

## Issue ticket number and link

https://jira.support-ent.fr/browse/MAG-58

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [X] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [X] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/resources/public/ts/directives/linker/linker.directive.ts [hunk 1]
```diff
@@ -22,27 +43,131 @@ interface ILinkerScope extends IScope, ILinkerProps {
...
 
-    $onInit() {
+    async $onInit() {
+        await this.getApplications();
+        this.resetLinker();
+
+        this.display = false;
+
+        if (appPrefix) {
+            this.apps = _.map(this.apps, app => {
```
Comment (flmartineau) [inline src/main/resources/public/ts/directives/linker/linker.directive.ts:68]: t'as essaye plutôt `this.apps = this.apps.map(app => {` ?

File: src/main/resources/public/ts/directives/linker/linker.directive.ts [hunk 2]
```diff
@@ -22,27 +43,131 @@ interface ILinkerScope extends IScope, ILinkerProps {
...
+                app.displayName = idiom.translate(app.displayName);
+                return app;
+            });
+
+            const currentApp = _.find(this.apps, app => {
```
Comment (flmartineau) [inline src/main/resources/public/ts/directives/linker/linker.directive.ts:73]: `const currentApp = this.apps.find((app: metLeTypeIci) => {`

File: src/main/resources/public/ts/directives/linker/linker.directive.ts [hunk 3]
```diff
@@ -22,27 +43,131 @@ interface ILinkerScope extends IScope, ILinkerProps {
...
+        }
+
     }
 
-    isFormValid = (): boolean => {
-        return this.form && this.form.url !== '';
+    resetLinker = (): void => {
+        this.searching = false;
+        this.search = {
+            application: this.apps[0],
```
Comment (flmartineau) [inline src/main/resources/public/ts/directives/linker/linker.directive.ts:85]: `this.apps && this.apps.length > 0 ? this.apps[0]: null`

File: src/main/resources/public/ts/directives/linker/linker.directive.ts [hunk 4]
```diff
@@ -22,27 +43,131 @@ interface ILinkerScope extends IScope, ILinkerProps {
...
+            const result = behaviour.loadResources(this.searchApplication)
+            if (result && result.then) {
+                result.then(() => {
+                    this.searchApplication();
+                });
+            }
+        }
+    }
+
+    searchResources = (appBehaviour): void => {
```
Comment (flmartineau) [inline src/main/resources/public/ts/directives/linker/linker.directive.ts:133]: typage de appBehaviour ? (et typages dans les callbacks aussi)

File: src/main/resources/public/ts/models/linker.model.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+export interface ILinkerParams {
+    id?: string,
+    appPrefix?: string,
+    link?: string,
+    blank?: boolean,
+    target?: string,
+    title?: string
+}
+
+export interface Resource {
```
Comment (flmartineau) [inline src/main/resources/public/ts/models/linker.model.ts:15]: IResource ?

File: src/main/resources/public/ts/models/linker.model.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    _id?: string,
+    title?: string,
+    path?: string
+}
+
+export interface App {
```
Comment (flmartineau) [inline src/main/resources/public/ts/models/linker.model.ts:21]: IApp ?

File: src/main/resources/public/ts/services/linker.service.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+    getResourcesApplications(): Promise<App[]>;
+
+}
+
+export const linkerService: ILinkerService = {
+
+    getResourcesApplications: async (): Promise<App[]> => {
+        return http.get('/resources-applications').then((res: AxiosResponse) => {
+            return this.apps = _.filter(model.me.apps, function (app) {
```
Comment (flmartineau) [inline src/main/resources/public/ts/services/linker.service.ts:16]: essaye `this.apps = model.me.apps.filter((app: TYPEICI) => `

File: src/main/resources/public/ts/services/linker.service.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+                return _.find(res.data, function (match) {
```
Comment (flmartineau) [inline src/main/resources/public/ts/services/linker.service.ts:17]: res.data.find((match: _type_) => {

**Review by flmartineau — APPROVED** [review]

**Review by flmartineau — APPROVED** [review]

File: src/main/resources/public/ts/directives/linker/linker.directive.ts [hunk 9]
```diff
@@ -22,27 +43,138 @@ interface ILinkerScope extends IScope, ILinkerProps {
...
+
+    searchApplication = (): void => {
+        let split: string[] = this.search.application.address.split('/');
+        this.params.appPrefix = split[split.length - 1];
+        Behaviours.loadBehaviours(this.params.appPrefix, this.searchResources);
+    };
+
+
+    loadBehaviour = (): void => {
+        const behaviour = Behaviours.applicationsBehaviours[this.params.appPrefix];
```
Comment (sjollois) [inline src/main/resources/public/ts/directives/linker/linker.directive.ts:119]: typage

File: src/main/resources/public/ts/directives/linker/linker.directive.ts [hunk 10]
```diff
@@ -22,27 +43,138 @@ interface ILinkerScope extends IScope, ILinkerProps {
...
+        this.addResource = behaviour.create;
+        if (behaviour.loadResourcesWithFilter) {
+            this.searchApplication();
+        } else if (behaviour.loadResources) {
+            const result = behaviour.loadResources(this.searchApplication)
```
Comment (sjollois) [inline src/main/resources/public/ts/directives/linker/linker.directive.ts:124]: typage

File: src/main/resources/public/ts/directives/linker/linker.html [hunk 11]
```diff
@@ -1,35 +1,84 @@
...
-                        data-ng-click="vm.submitLink()"
-                        data-ng-disabled="!vm.isFormValid()">
-                    <i18n>magneto.create.link</i18n>
-                </button>
-                <button class="right-magnet cancel"
-                        data-ng-click="vm.closeForm()">
-                    <i18n>magneto.cancel</i18n>
-                </button>
+<lightbox show="vm.display" on-close="vm.closeForm()">
+    <h2>Créer un lien</h2>
```
Comment (sjollois) [inline src/main/resources/public/ts/directives/linker/linker.html:2]: i18n

File: src/main/resources/public/ts/directives/linker/linker.html [hunk 12]
```diff
@@ -1,35 +1,84 @@
...
+        <header ng-click="vm.externalLink = true" ng-class="{ selected: vm.externalLink }">
+            <i18n>magneto.link.external</i18n>
+        </header>
+
+        <section ng-if="vm.externalLink">
+            <div class="row">
+                <label class="four cell zero-mobile">
+                    <i18n>magneto.link.address</i18n>
+                </label>
+                <input class="four cell twelve-mobile" type="text" placeholder="http://adresse-du-site.com"
```
Comment (sjollois) [inline src/main/resources/public/ts/directives/linker/linker.html:16]: unplaceholder un peu plus général (pas trop français)

**Review by sjollois — APPROVED** [review]

File: src/main/resources/public/ts/directives/linker/linker.html (hunk 12 above)
Comment (qpic-cgi) [reply src/main/resources/public/ts/directives/linker/linker.html:16]: i18n-placeholder maybe ?

**Review by qpic-cgi — APPROVED** [review]

**Pull request merged**